tfm: Configure TFM via Kconfig instead of Cmake

oyvindronningstad opened the pull request:
Add Kconfig options that will be used by the module to call the function
with the desired parameters. Refactor the tfm_integration samples and
the supported boards.

Update west.yml to bring in Cmake changes that use the new KConfigs.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

TFM PR: https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/21

oyvindronningstad committed:
tfm: Configure TFM via Kconfig instead of Cmake

Add Kconfig options that will be used by the module
to call the function with the desired parameters.
Refactor the tfm_integration samples and
the supported boards.

Update west.yml to bring in Cmake changes that use the new KConfigs.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

agansari commented on modules/Kconfig.tfm:
```diff
@@ -3,6 +3,20 @@
+	default "nordic_nrf/nrf9160dk_nrf9160" if BOARD_NRF9160DK_NRF9160NS
+	default "nordic_nrf/nrf5340dk_nrf5340_cpuapp" if BOARD_NRF5340DK_NRF5340_CPUAPPNS
+	default "nordic_nrf/nrf5340pdk_nrf5340_cpuapp" if BOARD_NRF5340PDK_NRF5340_CPUAPPNS
+	default "nxp/lpcxpresso55s69" if BOARD_LPCXPRESSO55S69_CPU0 || BOARD_LPCXPRESSO55S69_CPU1
```
CPU1 doesn't have trusted firmware extension

oyvindronningstad replied:
Thanks, removed CPU1.

erwango commented on modules/Kconfig.tfm:
```diff
@@ -3,6 +3,20 @@
+
+config TFM_BOARD
+	string
+	default "nordic_nrf/nrf9160dk_nrf9160" if BOARD_NRF9160DK_NRF9160NS
+	default "nordic_nrf/nrf5340dk_nrf5340_cpuapp" if BOARD_NRF5340DK_NRF5340_CPUAPPNS
+	default "nordic_nrf/nrf5340pdk_nrf5340_cpuapp" if BOARD_NRF5340PDK_NRF5340_CPUAPPNS
+	default "nxp/lpcxpresso55s69" if BOARD_LPCXPRESSO55S69_CPU0
+	default "mps2/an521" if BOARD_MPS2_AN521
+	default "stm/nucleo_l552ze_q" if BOARD_NUCLEO_L552ZE_Q
+	default "musca_b1" if BOARD_MUSCA_B1
```
I would rather define this symbol in respective boards.
I don't think we want this file to be updated each time a board supporting TFM is added.

oyvindronningstad replied:
I get your point. IMO there are a few good reasons to do it here though:
1) It's more concise.
2) It gives a better overview of what boards are available.
3) This is conceptually a property of TFM as much as it is a property of the board, since for CONFIG_TFM_BOARD to be valid, the port for the board must be present in TFM.

How would you envision defining it in the board files?

microbuilder replied:
I agree, @erwango, that my first instinct is to keep this at the board level as well, but points 2 and 3 from @oyvindronningstad  make a lot of sense. This makes it less likely any boards are overlooked when dealing with upstream TF-M changes, and does help identify what is currently supported in terms of HW.

ioannisg replied:
Regarding {3}, I think that this is basically and above all, a property of Tfm, not Zephyr. It is the list of target names of HW supported by tfm. It is not a Board property, in the sense that a Board can not support building with tfm unless the tfm module supports the board ;)

Regarding (2), I think that's also a good point, because we really want to keep module specific info in as few places as possible, in the tree, instead of distributing them sliver the place. That's because tfm is an external module and not a regular zephyr subsystem (in which case I would argue against)

erwango replied:
I'm not sure to agree on any of these points as they would also be true for any subsystem (net, usb, ...).
But applying these rules would get board references everywhere and it would make it things complex when adding boards.

This being said, I can see that I'm clearly the minority here, so I'll remove my nack. Let's see in long term how things evolve. Thanks for having taken the time to discuss my point.

ioannisg replied:
Sure @erwango , and of course, this is a valid point to discuss, and I appreciate your feedback a lot. This allows me to elaborate :)

**In my (I admit: subjective) point of view**, the difference between a subsystem and TFM is that the subsystem is part of Zephyr, maintained by Zephyr community members, while TFM is a fully external project, which Zephyr does not have control over and does not depend on at all (contrary e.g. to vendor HALs, which Zephyr _does_ depend on).

To give an example of the impact of our decision: if TFM decides to change the platform target names (which they actually _did_ 2 months ago) we would need to apply fixes inside the Zephyr board definitions. This is sort of counter-intuitive; why would Zephyr board definitions depend on an external project? Zephyr Cortex-M33 boards can be used without TFM at all, so I don't clearly see why any TFM-specific info should be present in generic board def-konfig files.

TFM target names are TFM-specific info and would rather reside on the module information. So if the target names change again in the future (hope, not), we would need to modify a single module Kconfig file, which is, sort of, the shim layer between the module and Zephyr.

That should apply to every external module - not only TFM.

Note also that this is in accordance with what we agreed in TSC regarding dealing with modules - minimize the "interface" between module and Zephyr and ideally make it reside at the "Zephyr" side instead of "module side".

tejlmand replied:
Sorry to be a little late on this, but I agree with @erwango .
This belongs with boards.

If users create custom board in a Zephyr module, they will usually base a new board on existing board conf.

Like creating a `nrf9160_inblue21`.
I don't like to see a huge list of boards in the Kconfig.tfm file. as this PR proposes.

To me it seems that with little effort it is possible to use the TFM_BOARD nrf9160dk_nrf9160 board together with Zephyr board `nrf9160_inblue21` and in that way get TFM support.

The only place I see board (and not SoC) specific settings is in: https://github.com/zephyrproject-rtos/trusted-firmware-m/blob/master/trusted-firmware-m/platform/ext/target/nordic_nrf/nrf9160dk_nrf9160/nrf_board.h

and that file seems to be only used for test purposes.

Updating the `<board>/Kconfig.defconfig` with compatible TFM board seems more logical, and it allows users with custom boards in custom Zephyr modules to be able to use TFM without having to modify `Kconfig.tfm` in Zephyr repo.

With TFM, one could probably set `TFM_PLATFORM` to `../../../../../some/escaped/path/my_board`.
Of course the best will be if TFM makes it possible to have custom boards outside `TFM` repo in a proper way.
So that one could do `set(TFM_PLATFORM <my_board_dir>)` 

In which case, Zephyr itself would allow for implementing a TFM valid board, even if it is not (yet) part of TFM upstream.

>  1. It's more concise.

Compared to what ?

> 2. It gives a better overview of what boards are available.

No, this give a better overview:
https://github.com/zephyrproject-rtos/trusted-firmware-m/tree/master/trusted-firmware-m/platform/ext/target

But what about future if TFM starts support TFM boards outside the TFM repo ?
Like I describe above ?

> 3.   This is conceptually a property of TFM as much as it is a property of the board, since for CONFIG_TFM_BOARD to be valid, the port for the board must be present in TFM.

Well, when TFM allows for custom boards, which I hope they will, then being able to specify the TFM Board together with a Zephyr board would make sense for a Zephyr downstream user.

oyvindronningstad replied:
> and that file seems to be only used for test purposes.

The difference between boards is not big, but it is used e.g. for UART printout, which is not test-specific.

> Compared to what ?

1 line per board instead of 2-5 lines (distributed across different files).

> No, this give a better overview:
> https://github.com/zephyrproject-rtos/trusted-firmware-m/tree/master/trusted-firmware-m/platform/ext/target

Not for the user who sees Kconfig complaining that TFM_BOARD is an empty string.

> But what about future if TFM starts support TFM boards outside the TFM repo ?

This is a valid point, but not until we actually have it IMO. At which point we can refactor this.

tejlmand replied:
> > Compared to what ?
> 
> 1 line per board instead of 2-5 lines (distributed across different files).
> 

That is a very bad argument, imho.
That is like saying that: 
https://github.com/zephyrproject-rtos/zephyr/blob/5a58ad508cb37c151271b4b6b6b3dfe589a09be8/drivers/spi/Kconfig.stm32#L15-L16
```
config SPI_STM32_INTERRUPT
	bool "STM32 MCU SPI Interrupt Support"
        default y if BOARD_NUCLEO_F091RC
        default y if BOARD_96B_NEONKEY
        default y if BOARD_96B_CARBON
        default y if BOARD_NUCLEO_L496ZG
        default y if BOARD_STM32F469I_DISCO
        default y if BOARD_STM32F769I_DISCO
        default y if BOARD_STM32F072B_DISCO
        default y if BOARD_NUCLEO_L476RG
        default y if BOARD_NUCLEO_L073RZ
        default y if BOARD_STM32F723E_DISCO
        default y if BOARD_NUCLEO_F401RE
        default y if BOARD_STM32MP157C_DK2
        default y if BOARD_NUCLEO_G431RB
        default y if BOARD_NUCLEO_L011K4
        default y if BOARD_NUCLEO_F411RE
        default y if BOARD_NUCLEO_F070RB
        default y if BOARD_NUCLEO_L053R8
        default y if BOARD_STM32L496G_DISCO
        default y if BOARD_SENSORTILE_BOX
        default y if BOARD_OLIMEXINO_STM32
        default y if BOARD_NUCLEO_G474RE
        default y if BOARD_DISCO_L475_IOT1
        default y if BOARD_BLACKPILL_F411CE
        default y if BOARD_NUCLEO_L432KC
        default y if BOARD_BLACK_F407ZG_PRO
        default y if BOARD_NUCLEO_L4R5ZI
        default y if BOARD_NUCLEO_L031K6
        default y if BOARD_NUCLEO_F334R8
        default y if BOARD_96B_STM32_SENSOR_MEZ
        default y if BOARD_BLACK_F407VE
        default y if BOARD_MIKROE_MINI_M4_FOR_STM32
        default y if BOARD_NUCLEO_F446RE
        default y if BOARD_B_L4S5I_IOT01A
```
would be better, instead of current design with:
https://github.com/zephyrproject-rtos/zephyr/blob/5a58ad508cb37c151271b4b6b6b3dfe589a09be8/boards/arm/nucleo_f091rc/Kconfig.defconfig#L11-L13

https://github.com/zephyrproject-rtos/zephyr/blob/5a58ad508cb37c151271b4b6b6b3dfe589a09be8/boards/arm/96b_neonkey/Kconfig.defconfig#L11-L13

and so on.

Just because you would be able to see all boards that supports SPI Interrupt for STM in a single file.

The TFM setting is a board setting, and by having it with board definition then a user can easily see what this board configures/supports. Just like with any other board configuration.

And a board maintainer can update / create a new board, without needing to update a lot of different Kconfig files throughout multiple folders / repos.

This PR is the first time that config settings inside Zephyr starts to default to a given value based on the board.
A default value based on the board only exists in boards or soc folder.

(There are a couple of exceptions in drivers for QEMU/native posix boards)

tejlmand replied:
So having this in `<boards>/Kconfig.defconfig` would be consistent with everywhere else in Zephyr.
And consistency is king.

tejlmand replied:
> 
> Not for the user who sees Kconfig complaining that TFM_BOARD is an empty string.

Then don't depend on the string, but have a `TFM_SUPPORT` that boards supporting TFM can set.
Like:
```
config TFM_SUPPORT
        bool
```

and users will see:
```
warning: BUILD_WITH_TFM (defined at modules/Kconfig.tfm:23) was assigned the value 'y' but got the
value 'n'. Check these unsatisfied dependencies: TRUSTED_EXECUTION_NONSECURE (=n), TFM_SUPPORT (=n).
```

tejlmand replied:
I have looked closer at the code.
As we both agree, then TFM does not allow for custom tfm platform folders at the moment.
Thus, there is no reason that the Kconfig contain strings, as the TFM_PLATFORM folder is a build system structure, not a config option.

Thus I propose to change into:
```
# Kconfig.tfm
config TFM_BOARD
        bool

config TFM_BOARD_NRF9160DK_NRF9160
        select TFM_BOARD
        bool

config TFM_BOARD_LPCXPRESSO55S69
        select TFM_BOARD
        bool

.... continue with rest of boards

menuconfig BUILD_WITH_TFM
        bool "Build with TF-M as the Secure Execution Environment"
        ...
        depends on TRUSTED_EXECUTION_NONSECURE
        depends on TFM_BOARD
```

and in: https://github.com/zephyrproject-rtos/trusted-firmware-m/blob/master/CMakeLists.txt
```
  if(CONFIG_TFM_BOARD_NRF9160DK_NRF9160)
    set(TFM_BOARD_DIR nordic_nrf/nrf9160dk_nrf9160)
  elseif(CONFIG_TFM_BOARD_LPCXPRESSO55S69)
    set(TFM_BOARD_DIR nxp/lpcxpresso55s69)
  elseif(...)
    .... continue with rest of boards
  endif()
```

board maintainers can then do:
```
# <board>/Kconfig.defconfig
config TFM_BOARD_<TFM_BOARD>
        default y
```
and thus they do not need to know anything about internal CMake folder structure inside TFM repo, and thus removing @ioannisg concern if TFM restructure later.
The shim becomes the `TFM_BOARD_<name>` setting, and Zephyr decides that name, and Zephyr TFM glue code decides how that setting maps into TFM CMake build structure.

as example for `nrf9160dk_nrf9160/Kconfig.defconfig`:
```
if BOARD_NRF9160DK_NRF9160NS

config TFM_BOARD_NRF9160DK_NRF9160
        default y

```

That will follow existing board defconfig pattterns, and place the CMake build folder structure where it belongs, inside CMakeLists.txt.

erwango replied:
@ioannisg, @oyvindronningstad, I find that this latest proposal from @tejlmand is a good middleground.
I would be in favor of a new PR to move to this method

ioannisg replied:
I did not read the whole conversation but I lean towards accepting the latest proposal. I'll open a PR and we can finalize the discussion there ( @microbuilder included :) )

tejlmand replied:
> I did not read the whole conversation but I lean towards accepting the latest proposal. I'll open a PR and we can finalize the discussion there ( @microbuilder included :) )

@ioannisg did you forget about this ?

anangl commented on modules/Kconfig.tfm:
```diff
@@ -3,6 +3,20 @@
+	default "nordic_nrf/nrf5340dk_nrf5340_cpuapp" if BOARD_NRF5340DK_NRF5340_CPUAPPNS
+	default "nordic_nrf/nrf5340pdk_nrf5340_cpuapp" if BOARD_NRF5340PDK_NRF5340_CPUAPPNS
+	default "nxp/lpcxpresso55s69" if BOARD_LPCXPRESSO55S69_CPU0
+	default "mps2/an521" if BOARD_MPS2_AN521
```
The corresponding `board.cmake` line should be removed:
https://github.com/zephyrproject-rtos/zephyr/blob/d3ef29af44166eccb8620430c7b751fe5209cee1/boards/arm/mps2_an521/board.cmake#L4

oyvindronningstad replied:
Thanks, removed.

oyvindronningstad commented on boards/arm/nrf9160dk_nrf9160/board.cmake:
```diff
@@ -3,7 +3,6 @@
 # Set the corresponding TF-M target platform when building for the Non-Secure
 # version of the board (Application MCU).
```
Thanks, removed.

ioannisg reviewed (approved):
Looks OK to me. Note that it requires the PR from @agansari , to restore the isolation level of nucleo board. But we can merge both PRs at the same time.

@erwango @agansari could you please revisit your reviews?

ioannisg committed:
tf-m: Add help text in some un-documented TF-M Kconfig options

Add help text in some newly introduced and un-documented
TF-M Kconfig options: REGRESSION and IPC.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
Added a new commit with some help text for the new options that were introduced here.

ioannisg commented:
Removed DNM since the TFM module SHA is updated.

tejlmand commented on modules/Kconfig.tfm:
```diff
@@ -1,8 +1,23 @@
+
+config TFM_BOARD
+	string
+	default "nordic_nrf/nrf9160dk_nrf9160" if BOARD_NRF9160DK_NRF9160NS
```
Little late to the game, but why have all this knowledge here ?

This make users think they need to update this file in order to set `TFM_BOARD`.
For example, for the `nrf9160dk_nrf9160ns` I think it would be better to have this in:
https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/nrf9160dk_nrf9160/Kconfig.defconfig
```
if BOARD_NRF9160DK_NRF9160NS
...

config TFM_BOARD
        default "nordic_nrf/nrf9160dk_nrf9160"

endif
```

oyvindronningstad replied:
Discussion here: https://github.com/zephyrproject-rtos/zephyr/pull/30062#discussion_r524413356
